<a href="https://colab.research.google.com/github/hijera/NN_colab_hw/blob/main/%D0%94%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D0%B5%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_%D0%BA_%D0%BB%D0%B5%D0%BA%D1%86%D0%B8%D0%B8_%C2%AB%D0%A3%D0%BB%D1%83%D1%87%D1%88%D0%B5%D0%BD%D0%B8%D0%B5_%D0%BA%D0%B0%D1%87%D0%B5%D1%81%D1%82%D0%B2%D0%B0_%D0%BC%D0%BE%D0%B4%D0%B5%D0%BB%D0%B8%C2%BB_%D0%BE%D0%B1%D0%BD%D0%BE%D0%B2%D0%BB%D0%B5%D0%BD%D0%BD%D0%BE%D0%B5_(%D0%92%D0%BB%D0%B0%D1%81%D0%BE%D0%B2_%D0%90%D0%BD%D1%82%D0%BE%D0%BD).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Загрузка и преобразование данных

In [ ]:
import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder
from sklearn.pipeline import make_pipeline  # используем пайплайны для удобства
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_validate,GridSearchCV,RandomizedSearchCV

In [ ]:
import warnings
warnings.filterwarnings("ignore")

In [ ]:
rs=42

In [ ]:
data=pd.read_csv("./heart.csv")

In [ ]:
data

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             918 non-null    int64  
 1   Sex             918 non-null    object 
 2   ChestPainType   918 non-null    object 
 3   RestingBP       918 non-null    int64  
 4   Cholesterol     918 non-null    int64  
 5   FastingBS       918 non-null    int64  
 6   RestingECG      918 non-null    object 
 7   MaxHR           918 non-null    int64  
 8   ExerciseAngina  918 non-null    object 
 9   Oldpeak         918 non-null    float64
 10  ST_Slope        918 non-null    object 
 11  HeartDisease    918 non-null    int64  
dtypes: float64(1), int64(6), object(5)
memory usage: 86.2+ KB


In [ ]:
for name in data.keys().values.tolist():
  print(name,":")
  print(data[name].unique())
#print(data['Sex'].unique())
#print(data['ChestPainType'].unique())

Age :
[40 49 37 48 54 39 45 58 42 38 43 60 36 44 53 52 51 56 41 32 65 35 59 50
 47 31 46 57 55 63 66 34 33 61 29 62 28 30 74 68 72 64 69 67 73 70 77 75
 76 71]
Sex :
['M' 'F']
ChestPainType :
['ATA' 'NAP' 'ASY' 'TA']
RestingBP :
[140 160 130 138 150 120 110 136 115 100 124 113 125 145 112 132 118 170
 142 190 135 180 108 155 128 106  92 200 122  98 105 133  95  80 137 185
 165 126 152 116   0 144 154 134 104 139 131 141 178 146 158 123 102  96
 143 172 156 114 127 101 174  94 148 117 192 129 164]
Cholesterol :
[289 180 283 214 195 339 237 208 207 284 211 164 204 234 273 196 201 248
 267 223 184 288 215 209 260 468 188 518 167 224 172 186 254 306 250 177
 227 230 294 264 259 175 318 216 340 233 205 245 194 270 213 365 342 253
 277 202 297 225 246 412 265 182 218 268 163 529 100 206 238 139 263 291
 229 307 210 329 147  85 269 275 179 392 466 129 241 255 276 282 338 160
 156 272 240 393 161 228 292 388 166 247 331 341 243 279 198 249 168 603
 159 190 185 290 212 231 222 235 320 187 266 2

In [ ]:
#Преобразуем данные при помощи LabelEncoder и OneHotEncoder
y = data['HeartDisease']
X = data.copy()
del X['HeartDisease']
Xp=pd.get_dummies( X, columns = [ 'ST_Slope','RestingECG','ChestPainType' ] )
le1 = LabelEncoder()
Xp.ExerciseAngina = le1.fit_transform(Xp.ExerciseAngina)
le2 = LabelEncoder()
Xp.Sex=le2.fit_transform(Xp.Sex)
Xp

Age  Sex  RestingBP  Cholesterol  FastingBS  MaxHR  ExerciseAngina  \
0     40    1        140          289          0    172               0   
1     49    0        160          180          0    156               0   
2     37    1        130          283          0     98               0   
3     48    0        138          214          0    108               1   
4     54    1        150          195          0    122               0   
..   ...  ...        ...          ...        ...    ...             ...   
913   45    1        110          264          0    132               0   
914   68    1        144          193          1    141               0   
915   57    1        130          131          0    115               1   
916   57    0        130          236          0    174               0   
917   38    1        138          175          0    173               0   

     Oldpeak  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  RestingECG_LVH  \
0        0.0              0              0            1               0   
1        1.0              0              1            0               0   
2        0.0              0              0            1               0   
3        1.5              0              1            0               0   
4        0.0              0              0            1               0   
..       ...            ...            ...          ...             ...   
913      1.2              0              1            0               0   
914      3.4              0              1            0               0   
915      1.2              0              1            0               0   
916      0.0              0              1            0               1   
917      0.0              0              0            1               0   

     RestingECG_Normal  RestingECG_ST  ChestPainType_ASY  ChestPainType_ATA  \
0                    1              0                  0                  1   
1                    1              0                  0                  0   
2                    0              1                  0                  1   
3                    1              0                  1                  0   
4                    1              0                  0                  0   
..                 ...            ...                ...                ...   
913                  1              0                  0                  0   
914                  1              0                  1                  0   
915                  1              0                  1                  0   
916                  0              0                  0                  1   
917                  1              0                  0                  0   

     ChestPainType_NAP  ChestPainType_TA  
0                    0                 0  
1                    1                 0  
2                    0                 0  
3                    0                 0  
4                    1                 0  
..                 ...               ...  
913                  0                 1  
914                  0                 0  
915                  0                 0  
916                  0                 0  
917                  1                 0  

[918 rows x 18 columns]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(Xp, y, test_size=0.2,random_state=rs)

In [ ]:
y

0      0
1      1
2      0
3      1
4      0
      ..
913    1
914    1
915    1
916    1
917    0
Name: HeartDisease, Length: 918, dtype: int64

#Обучение модели

In [ ]:
score_params=['accuracy','recall','precision','f1']

In [ ]:
model=LogisticRegression(max_iter=1000)
model.fit(X_train,y_train)

LogisticRegression(max_iter=1000)

In [ ]:
cross_validate(model,X_train,y_train, cv=10, scoring=score_params)

{'fit_time': array([0.29142022, 0.23063827, 0.31595778, 0.27696109, 0.45567226,
        0.60529566, 0.24181437, 0.31538224, 0.34374475, 0.41901493]),
 'score_time': array([0.00958681, 0.12129378, 0.00959659, 0.02061772, 0.01964974,
        0.02102232, 0.0223341 , 0.01929331, 0.02920365, 0.01986384]),
 'test_accuracy': array([0.91891892, 0.87837838, 0.89189189, 0.86486486, 0.84931507,
        0.89041096, 0.76712329, 0.87671233, 0.87671233, 0.83561644]),
 'test_recall': array([0.925     , 0.875     , 0.875     , 0.90243902, 0.875     ,
        0.925     , 0.825     , 0.9       , 0.925     , 0.9       ]),
 'test_precision': array([0.925     , 0.8974359 , 0.92105263, 0.86046512, 0.85365854,
        0.88095238, 0.76744186, 0.87804878, 0.86046512, 0.81818182]),
 'test_f1': array([0.925     , 0.88607595, 0.8974359 , 0.88095238, 0.86419753,
        0.90243902, 0.79518072, 0.88888889, 0.89156627, 0.85714286])}

# Оптимизация параметров

In [ ]:
params=dict(max_iter=[500,1000,2000,5000],
                                      penalty=['l1','l2','elasticnet'],
                                             solver=['lbfgs', 'liblinear', 'newton-cg']
                                      )
grid=GridSearchCV( LogisticRegression(random_state=rs),params,
                   n_jobs=None,
                   cv=10, scoring='accuracy')
grid.fit(X_train,y_train)

GridSearchCV(cv=10, estimator=LogisticRegression(random_state=42),
             param_grid={'max_iter': [500, 1000, 2000, 5000],
                         'penalty': ['l1', 'l2', 'elasticnet'],
                         'solver': ['lbfgs', 'liblinear', 'newton-cg']},
             scoring='accuracy')

In [ ]:
cross_validate(grid,X_train,y_train, cv=10, scoring=score_params)

{'fit_time': array([16.10476351, 13.12163854, 13.72741175, 12.99285507, 15.03332496,
        14.27959847,  9.96191669, 12.55823183, 12.73908281, 15.79387426]),
 'score_time': array([0.00959396, 0.0093298 , 0.01311445, 0.00925612, 0.00934863,
        0.00958371, 0.00924969, 0.00903916, 0.00918651, 0.00959516]),
 'test_accuracy': array([0.93243243, 0.87837838, 0.89189189, 0.86486486, 0.84931507,
        0.89041096, 0.76712329, 0.84931507, 0.89041096, 0.83561644]),
 'test_recall': array([0.925     , 0.875     , 0.875     , 0.90243902, 0.875     ,
        0.925     , 0.825     , 0.85      , 0.95      , 0.9       ]),
 'test_precision': array([0.94871795, 0.8974359 , 0.92105263, 0.86046512, 0.85365854,
        0.88095238, 0.76744186, 0.87179487, 0.86363636, 0.81818182]),
 'test_f1': array([0.93670886, 0.88607595, 0.8974359 , 0.88095238, 0.86419753,
        0.90243902, 0.79518072, 0.86075949, 0.9047619 , 0.85714286])}

In [ ]:
print(grid.best_score_)

print(grid.best_params_)

print(grid.best_estimator_)

0.8677156608663458
{'max_iter': 500, 'penalty': 'l2', 'solver': 'liblinear'}
LogisticRegression(max_iter=500, random_state=42, solver='liblinear')


In [ ]:
rcv=RandomizedSearchCV(LogisticRegression(random_state=rs),params,cv=10,verbose=1,n_iter=30,scoring='accuracy')
rcv.fit(X_train,y_train)

Fitting 10 folds for each of 30 candidates, totalling 300 fits


RandomizedSearchCV(cv=10, estimator=LogisticRegression(random_state=42),
                   n_iter=30,
                   param_distributions={'max_iter': [500, 1000, 2000, 5000],
                                        'penalty': ['l1', 'l2', 'elasticnet'],
                                        'solver': ['lbfgs', 'liblinear',
                                                   'newton-cg']},
                   scoring='accuracy', verbose=1)

In [ ]:
cross_validate(rcv,X_train,y_train, cv=10, scoring=score_params)

Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits
Fitting 10 folds for each of 30 candidates, totalling 300 fits


{'fit_time': array([11.04265499, 11.40712166,  9.70151567, 11.48637152, 11.89481091,
        13.92290449,  9.27486062,  9.88230419, 11.26087141, 11.32949996]),
 'score_time': array([0.00636053, 0.01432967, 0.00933313, 0.00897694, 0.00922871,
        0.00921607, 0.00934243, 0.00935316, 0.01962042, 0.00934863]),
 'test_accuracy': array([0.93243243, 0.87837838, 0.89189189, 0.86486486, 0.84931507,
        0.89041096, 0.76712329, 0.84931507, 0.89041096, 0.83561644]),
 'test_recall': array([0.925     , 0.875     , 0.875     , 0.90243902, 0.875     ,
        0.925     , 0.825     , 0.85      , 0.95      , 0.9       ]),
 'test_precision': array([0.94871795, 0.8974359 , 0.92105263, 0.86046512, 0.85365854,
        0.88095238, 0.76744186, 0.87179487, 0.86363636, 0.81818182]),
 'test_f1': array([0.93670886, 0.88607595, 0.8974359 , 0.88095238, 0.86419753,
        0.90243902, 0.79518072, 0.86075949, 0.9047619 , 0.85714286])}

In [ ]:
print(rcv.best_score_)

print(rcv.best_params_)

print(rcv.best_estimator_)

0.8677156608663458
{'solver': 'liblinear', 'penalty': 'l2', 'max_iter': 500}
LogisticRegression(max_iter=500, random_state=42, solver='liblinear')


#Использование различных моделей

In [ ]:
from sklearn.svm import SVC
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
#Подготоваливаем словарь моделей
algos = [
  SVC(random_state=rs),
  DecisionTreeClassifier(random_state=rs),
  RandomForestClassifier(random_state=rs)
]
a_params= [
    dict(kernel=['linear', 'poly', 'rbf', 'sigmoid'],max_iter=[500,1000,2000,3000]),
    dict(min_samples_leaf=[1,5,10,30],max_depth=[5,10,20]),
    dict(min_samples_leaf=[1,5,10,30],max_depth=[5,10,20])
]

In [ ]:
i=0
rr=[None,None,None]
for algo in algos:
  print(a_params[i])
  rr[i]=RandomizedSearchCV(algo,a_params[i],cv=10,verbose=1,n_iter=30,scoring='accuracy')
  rr[i].fit(X_train,y_train)
  print(cross_validate(rr[i],X_train,y_train, cv=10, scoring=score_params))
  print(rr[i].best_score_)
  print(rr[i].best_params_)
  print(rr[i].best_estimator_)
  i+=1

{'kernel': ['linear', 'poly', 'rbf', 'sigmoid'], 'max_iter': [500, 1000, 2000, 3000]}
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
Fitting 10 folds for each of 16 candidates, totalling 160 fits
{'fit_time': array([3.46231604, 3.44262791, 4.15353131, 3.59070015, 3.45855522,
       3.4251833 , 4.57402372, 3.443923  , 3.42640424, 6.1865108 ]), 'score_time': array([0.00852108, 0.00966311, 0.01212978, 0.00730729, 0.0

#Вывод

RandomizedSearchCV и GridSearchCV выдали одинаковую точность при разных параметрах. Использование различных моделей позволило слегка улучшить результат - наилучший получился при использовании RandomForest (accuracy = 0.87). Примерно такой же результат у нас получился при ансамблировании ( https://colab.research.google.com/drive/1lqsjO88ouYcPm1NvALpkyKzP7Fqs2WgE?usp=sharing )